iot/end_to_end_example/server/src/server.kt

```diff
@@ -0,0 +1,126 @@
+// Copyright 2018 Google Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+package com.google.iot
+
+import com.google.cloud.pubsub.v1.AckReplyConsumer
+import com.google.cloud.pubsub.v1.MessageReceiver
+import com.google.cloud.pubsub.v1.Subscriber
+
+import com.google.pubsub.v1.ProjectSubscriptionName
+import com.google.pubsub.v1.PubsubMessage
+
+import com.google.cloud.iot.v1.DeviceManagerClient
+import com.google.cloud.iot.v1.DeviceName
+import com.google.cloud.iot.v1.ModifyCloudToDeviceConfigRequest
+
+import com.google.protobuf.ByteString
+
+import java.io.IOException
+import java.util.concurrent.LinkedBlockingDeque
+
+import com.beust.klaxon.Klaxon
+import com.beust.klaxon.JsonObject
+
+private val iot = DeviceManagerClient.create()
+
+data class IotData(val temperature: Int)
+
+// The main loop. Consumes messages from the Pub/Sub subscription.
+fun main(args: Array<String>) {
+    if (args.size < 2) {
+        error("Usage: java -jar server.jar PROJECT_ID SUBSCRIPTION_ID\n")
+    }
+
+    val projectId = args[0]
+    val subscriptionId = args[1]
+    val messages = LinkedBlockingDeque<PubsubMessage>()
+
+    class MessageReceiverExample : MessageReceiver {
+        override fun receiveMessage(message: PubsubMessage, consumer: AckReplyConsumer) {
+            messages.offer(message)
+            consumer.ack()
+        }
+    }
+
+    val subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId)
+
+    val subscriber = Subscriber
+        .newBuilder(subscriptionName, MessageReceiverExample()).build()
+
+    try {
+        // Create a subscriber bound to the asynchronous message receiver
+        subscriber.startAsync().awaitRunning()
+
+        // Continue to listen to messages
+        println("Listening to messages on $subscriptionId:")
+        while (true) {
```

Comment:
why is there a while loop here? the MessageReceiver should handle the messages; not sure why there is a need to push to a queue etc?

Reply:
This is in the Java sample as well. My understanding is it's just to keep the program running indefinitely. Is there a better way to go about that?

Reply:
it's possible to prevent the process for exiting w/o the `while(true)` here.

I think the message receiver pushing received messages onto a in-memory queue, and the while loop here is reading off of the queue and then handling it serially.  That the message receiver should be able to call `handle_pubsub_message` directly. The only reason I think this is done this way is to force the use of the main thread to handle the messages serially. Not sure if it's required though.